eclass/zig.eclass

```diff
@@ -0,0 +1,572 @@
+# @DESCRIPTION:
+# If enabled, eclass will add "--summary all --verbose" options to
+# "ezig build", so that it prints every command before executing,
+# and summarry tree at the end of step.  If not, will do nothing.
+# Enabled by default.  Set to OFF to disable these verbose messages.
+#
+# Note: this variable does not control other options starting with
+# "--verbose-", such as "--verbose-link" or "--verbose-cimport".  If
+# you need them, add them manually to ZBS_ARGS_EXTRA.
+: "${ZBS_VERBOSE:=ON}"
+
+# @ECLASS_VARIABLE: BUILD_DIR
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Directory where all "ezig build" calls will be proceeded.
+# Defaults to "${WORKDIR}/${P}-build" if not set.
+: "${BUILD_DIR:=${WORKDIR}/${P}-build}"
+
+# @ECLASS_VARIABLE: ZBS_ECLASS_DIR
+# @DESCRIPTION:
+# Directory where various files used by this eclass are stored.
+# They can be supplied by the ebuild or by eclass.
+# Currently, it's used only for Zig packages, which are stored in "p/"
+# subdirectory.
+# Defaults to "${WORKDIR}/zig-eclass" if not set.
+# Should be set before calling "zig_src_unpack" or
+# "zig_live_fetch".
+: "${ZBS_ECLASS_DIR:=${WORKDIR}/zig-eclass}"
+
+# @FUNCTION: zig_get_jobs
+# @DESCRIPTION:
+# Returns number of jobs from ZBS_ARGS_EXTRA or MAKEOPTS.
+# If there is none, defaults to number of available processing units.
+zig_get_jobs() {
+	local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
+	local default_jobs="$(get_nproc)"
+	local jobs="$(makeopts_jobs "${all_args}" "${default_jobs}")"
+
+	if [[ "${jobs}" == "0" ]]; then
+		# Zig build system does not allow "-j0", and does not have
+		# option for unlimited parallelism. Pass our default number
+		# of jobs here.
+		echo "${default_jobs}"
+	else
+		echo "${jobs}"
+	fi
+}
+
+# @FUNCTION: zig_init_base_args
+# @DESCRIPTION:
+# Stores basic args for future "ezig build" calls in ZBS_ARGS_BASE.
+# Package manager option is managed by "zig_src_prepare",
+# ebuild and user options are added by "zig_src_configure".
+#
+# This function is used by "zig_pkg_setup", and it is neccessary
+# that args are available as early as possible, so that ebuilds
+# could use them in steps like "src_unpack" if neccessary, while
+# still having verbosity and amount of jobs from user respected.
+#
+#
+# TODO: currently this function enables "--search-prefix" (1) and
+# "--sysroot" (2) only when cross-compiling, should be fixed to
+# unconditionally enabling it.
+#
+# For solving (1) this patch should be reworked and upstreamed:
+# https://paste.sr.ht/~bratishkaerik/2ddffe2bf0f8f9d6dfb60403c2e9560334edaa88
+#
+# (2)
+# "--sysroot" should be passed together with "--search-prefix" above,
+# if we pass only "--sysroot" it gives these errors:
+# @CODE
+# error: unable to find dynamic system library 'zstd' using strategy 'paths_first'. searched paths: none
+# @CODE
+zig_init_base_args() {
+	[[ "${ZBS_ARGS_BASE}" ]] && return
+
+	local crt_dir="${ESYSROOT}/usr/"
+	if [[ "${ZIG_TARGET}" == *musl* ]]; then
+		crt_dir+="lib/"
```

Comment:
Hmm, doesn't `get_libdir` return the correct value on musl?

Reply:
It is specifically for `crt1.o` and similar files, which are located in `/usr/lib/` and not in `/usr/lib64` on musl:
```sh
# The directory that contains `crt1.o` or `crt2.o`.
# On POSIX, can be found with `cc -print-file-name=crt1.o`.

$ gcc -print-file-name=crt1.o
/usr/lib/gcc/x86_64-pc-linux-musl/13/../../../../lib/crt1.o

$ file /usr/lib{,64}/crt1.o
/usr/lib/crt1.o:   ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
/usr/lib64/crt1.o: cannot open `/usr/lib64/crt1.o' (No such file or directory)
```

Reply:
Ah, okay. I suppose asking the C compiler to print the path would be too much?

Reply:
> Ah, okay. I suppose asking the C compiler to print the path would be too much?

Sorry, I removed comment to not mislead, maybe I'm wrong. I just checked and `/usr/lib64` doesn't even exist on amd64 musl, so `get_libdir` should report correct value here. Maybe I mistook it with other musl arches, like aarch64 musl, but there was a reason 100%.

Reply:
Well, my initial thought was "but musl uses `lib` only", so I was kinda surprised you told me it uses lib64. I've just checked, and according to docs, they don't support libNN at all.

Reply:
b6c860c3e8de3b2d2df6f3468df72104b72aafde